polygeist/lib/Dialect/Polygeist/Analysis/UniformityAnalysis.cpp

```diff
@@ -75,93 +76,264 @@ void Uniformity::print(raw_ostream &os) const { os << *this; }
+
+  for (const auto &EI : effects) {
+    Value val = EI.getValue();
     if (!val) {
       // Memory effect on anything other than a value: conservatively assume
       // the result(s) uniformity is unknown.
       LLVM_DEBUG(llvm::dbgs() << "Memory Effect on non-values found\n");
-      for (UniformityLattice *result : results)
-        propagateIfChanged(result, result->join(Uniformity::getUnknown()));
+      return propagateAllIfChanged(results, Uniformity::getUnknown());
+    }
+
+    if (!isa<MemoryEffects::Read>(EI.getEffect()))
+      continue;
+
+    // Get the reaching def. and potential reaching def. of the value 'val' and
+    // analyze them to determine its uniformity.
+    const ReachingDefinition *rdef =
+        internalSolver.lookupState<ReachingDefinition>(op);
+    if (!rdef) {
+      LLVM_DEBUG(llvm::dbgs() << "Unable to find reaching definition\n");
+      return propagateAllIfChanged(op->getResults(), Uniformity::getUnknown());
+    }
+    LLVM_DEBUG(llvm::dbgs().indent(2) << "rdef:\n" << *rdef << "\n");
+
+    using ModifiersTy = ReachingDefinition::ModifiersTy;
+    auto merge = [](std::optional<ModifiersTy> mods,
+                    std::optional<ModifiersTy> pMods) {
+      if (!mods && !pMods)
+        return mods;
+      if (mods && !pMods)
+        return mods;
+      if (pMods && !mods)
+        return pMods;
+      mods->merge(*pMods);
+      return mods;
+    };
+
+    // Merge mods and pMods together.
+    std::optional<ModifiersTy> mods =
+        merge(rdef->getModifiers(val), rdef->getPotentialModifiers(val));
+    if (!mods)
+      continue;
+
+    // If any of mods/pMods are dominated by a branch with a condition that is
+    // unknown/non-uniform the loaded value has the same uniformity.
+    SmallVector<Value> branchConditions = collectBranchConditions(*mods);
+    if (anyOfUniformityIs(branchConditions, Uniformity::Kind::Unknown)) {
+      LLVM_DEBUG(llvm::dbgs().indent(2)
+                 << "Branch condition has unknown uniformity\n");
+      return propagateAllIfChanged(op->getResults(), Uniformity::getUnknown());
+    }
+    if (anyOfUniformityIs(branchConditions, Uniformity::Kind::NonUniform)) {
+      LLVM_DEBUG(llvm::dbgs().indent(2) << "Branch condition non-uniform\n");
+      return propagateAllIfChanged(op->getResults(), Uniformity::getUnknown());
+    }
+
+    // If we can't yet compute the mods/pMods operands uniformity, bail out.
+    if (!canComputeUniformity(*mods)) {
+      LLVM_DEBUG(llvm::dbgs().indent(2)
+                 << "Reaching def operand(s) uniformity not yet initialized\n");
       return;
     }
 
-    TypeSwitch<MemoryEffects::Effect *>(effect.getEffect())
-        .Case<MemoryEffects::Read>([](auto) {
-          // A read operation yields uniform result(s) iff the (potentially)
-          // reaching definitions of all its operands are uniform.
-          assert(false && "TODO");
+    // If any modifiers or potential modifiers of the value loaded store a value
```

Comment:
Do we also need to check the address operand of stores for uniformity? Could such a store still be a potential reaching definition and therefore introduce non-uniformity if some threads write to the same location as this load and some don't?

Reply:
Correct, we need to check that the memref used by a load operation (e.g. the address operand) is [uniform|non-uniform|unknonw]. Here is an example (that I am adding to the lit tests) showing what happens when a branch with a non-uniform condition allocates 2 different memory locations (alloca1, alloca2). Because the branch is divergent, the value yielded by the `scf.if` is non-uniform. The load that uses that value ought to also be non-uniform. 

This is working with the current implementation. 

```
  // COM: the branch condition is non-uniform, so the value yielded by the scf.if is also non-uniform.
  %c0_i64 = arith.constant 0 : i64
  %cond1 = arith.cmpi sgt, %tx, %c0_i64 : i64
  %alloca1 = scf.if %cond1 -> memref<10xi64> {
    %alloca1 = memref.alloca() : memref<10xi64>
    scf.yield %alloca1 : memref<10xi64>
  } else {
    %alloca2 = memref.alloca() : memref<10xi64>
    scf.yield %alloca2 : memref<10xi64>    
  }
  // CHECK: test2_load6, uniformity: non-uniform
  %load6 = memref.load %alloca1[%c0] { tag = "test2_load6" } : memref<10xi64>

Reply:
I think I had a more complicated scenario in mind, something like this: 
```
  %c0_i64 = arith.constant 0 : i64
  %alloca0 = memref.alloca() : memref<10xi64>
  %cond1 = arith.cmpi sgt, %tx, %c0_i64 : i64
  %alloca1 = scf.if %cond1 -> memref<10xi64> {
    scf.yield %alloca0 : memref<10xi64>
  } else {
    %alloca2 = memref.alloca() : memref<10xi64>
    scf.yield %alloca2 : memref<10xi64>    
  }
  memref.store %alloca1[%c0], %some_uniform_value

  // CHECK: test2_load6, uniformity: non-uniform
  %load6 = memref.load %alloca0[%c0] { tag = "test2_load6" } : memref<10xi64>
```

In this case, all operands of the `load` are uniform. However, the `store` might or might not write to the location, and whether it does depends on a non-uniform value (through the `scf.if`). 

Therefore, the `load` should still be determined as non-uniform.

Would the current implementation cover this scenario?

Reply:
@sommerlukas this commit (url)https://github.com/intel/llvm/pull/10098/commits/67d9a2682d398853ed2d8b3778e94648cc3d8ce3 addresses the test case you provided. Thanks for thinking of that test.